renku/core/models/workflow/parameters.py

```diff
@@ -17,106 +17,101 @@
 # limitations under the License.
 """Represents a workflow template."""
 
+import os
+import pathlib
+import urllib.parse
 import uuid
 
-from renku.core.models import jsonld as jsonld
+import attr
+from marshmallow import EXCLUDE
 
+from renku.core.models.calamus import JsonLDSchema, Nested, fields, rdfs, renku
+from renku.core.models.entities import CollectionSchema, EntitySchema
 
-@jsonld.s(
-    type=["renku:IOStream",],
-    context={
-        "renku": "https://swissdatasciencecenter.github.io/renku-ontology#",
-        "prov": "http://www.w3.org/ns/prov#",
-    },
-    cmp=False,
-)
+
+@attr.s(cmp=False,)
 class MappedIOStream(object):
     """Represents an IO stream (stdin, stdout, stderr)."""
 
-    _id = jsonld.ib(context="@id", kw_only=True)
-    _label = jsonld.ib(default=None, context="rdfs:label", kw_only=True)
+    client = attr.ib(default=None, kw_only=True)
+
+    _id = attr.ib(default=None, kw_only=True)
+    _label = attr.ib(default=None, kw_only=True)
 
     STREAMS = ["stdin", "stdout", "stderr"]
 
-    stream_type = jsonld.ib(
-        context={"@id": "renku:streamType", "@type": "http://www.w3.org/2001/XMLSchema#string",},
-        type=str,
-        kw_only=True,
-    )
+    stream_type = attr.ib(type=str, kw_only=True,)
 
-    @_id.default
     def default_id(self):
-        """Set default id."""
-        # TODO: make bnode ids nicer once this issue is in a release:
-        # https://github.com/RDFLib/rdflib/issues/888
-        # right now it's limited to a-zA-Z0-9 (-_ will work once it's fixed)
-        return "_:MappedIOStream-{}".format(str(uuid.uuid4())).replace("-", "")
+        """Generate an id for a mapped stream."""
+        host = "localhost"
+        if self.client:
+            host = self.client.remote.get("host") or host
```

Comment:
No need to branch, `get` internally supports default values.
```suggestion
            host = self.client.remote.get("host", host)
```

Reply:
Actually this doesn't work. In some cases, remote is 
```
> client.remote
{'host': None, 'name': None, 'owner': None}
```
so get returns `None` instead of the default value

Reply:
That smells like a serialization bug then. We should fix it in remote probably.

Reply:
Also, I believe that the property `remote` could be cached.

Reply:
it's what you get when you just renku init a repository locally with no remote. We could return `{}` instead but that's not any more or less correct than the current version. Ideally we'd return `None` if there is no remote, but then we'd have to guard that everywhere in code where it's used, so a separate ticket makes sense. --> https://github.com/SwissDataScienceCenter/renku-python/issues/1457